### ISSUE TITLE
can't insert â in xterm in modifyOtherKeys mode when <m-b> is mapped until an Ex command is run

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#5951]
**Describe the bug**

In xterm in modifyOtherKeys mode, some accented characters can't be inserted when some insert mode mappings are installed, until an Ex command is run.

**To Reproduce**

Open an xterm terminal.

Start Vim like this:

    vim -Nu NONE +'ino <m-b> xxx' +'startinsert'

Type `â`; on an azerty keyboard layout, it means pressing the dead key for the circumflex accent `^`, followed by `a`.  The result is that Vim inserts `xxx`.

Now, press Escape to get back to normal mode, `:` to enter the command-line, and Enter to run an empty command.  Finally, press `i` to enter insert-mode, and try again to insert `â`; this time, it works as expected; i.e. Vim inserts `â`.

**Expected behavior**

When typing `â`, Vim inserts `â` even if no Ex command has been run.

**Screenshots**

![gif](https://user-images.githubusercontent.com/8505073/79685516-bcc51200-8239-11ea-9e26-e7d022c5bc30.gif)

**Environment**

 - Vim version: 8.2 Included patches: 1-598
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(322)

**Additional context**

It doesn't matter which Ex command you run; however, you must run something; i.e. if you leave the command-line by pressing Escape instead of Enter, when you type `â`, Vim still uses the insert mode `<m-b>` mapping instead of inserting `â`.

---

Update: I've just found out that Vim correctly inserts `â` after you press `<M-b>` once.
But again, this is unexpected; Vim should insert `â` immediately, not after some other event has occurred.

**Comment by lacygoill** [comment]
As a workaround, I'm trying to make Vim run an empty Ex command just before entering insert mode for the first time:

    au InsertEnter * ++once call feedkeys("\<c-\>\<c-o>:\<cr>", 'nt')

But it doesn't work.  I've tried passing the `++nested` argument to `:au`, and the `L` flag to `feedkeys()`; it doesn't change anything.

OTOH, this works:

    au InsertEnter * ++once call feedkeys("\<c-\>\<c-o>:", 'nt')

But you still have to press Enter to validate the empty command-line.  Something must happen when you press Enter, which `feedkeys()` can't emulate, but I don't know what it is.

**Comment by brammool** [comment]
> As a workaround, I'm trying to make Vim enter the command-line at
> least once before entering insert mode:
> 
>     au InsertEnter * ++once call feedkeys("\<c-\>\<c-o>:\<cr>", 'nt')
> 
> But it doesn't work.  I've tried passed the `++nested` argument to
> `:au`, and the `L` flag to `feedkeys()`; it doesn't change anything.
> 
> OTOH, this works:
> 
>     au InsertEnter * ++once call feedkeys("\<c-\>\<c-o>:", 'nt')
> 
> But you still have to press Enter to validate the empty command-line.
> Something must happen when you press Enter, which `feedkeys()` can't
> emulate, but I don't know what it is.

Vim doesn't anything special when going to the command line and back.
Have you tried other things, such as CTRL-L?  Or clicking the mouse?
I suspect the terminal is in some strange mode.

-- 
EXPERIENCE - experience is a wonderful thing. It enables you to 
recognise a mistake when you make it again.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> Vim doesn't anything special when going to the command line and back.
> Have you tried other things, such as CTRL-L?  Or clicking the mouse?
> I suspect the terminal is in some strange mode.

Clicking the mouse doesn't help, however pressing `C-l` does help.  It seems that pressing any control character interactively fixes the issue.  But as soon as I use `feedkeys()` to automate the keypress, the fix stops working.

**Comment by lacygoill** [comment]
Ah you were right.  Entering the command-line is irrelevant.  What matters is `C-m`.  I wrote Enter but in reality I always press `C-m` instead of Enter.  If I press the Enter key, the issue persists; but `C-m` like all control keys (except `C-c`) fixes the issue.

**Comment by brammool** [comment]
> > Vim doesn't anything special when going to the command line and back.
> > Have you tried other things, such as CTRL-L?  Or clicking the mouse?
> > I suspect the terminal is in some strange mode.
> 
> Clicking the mouse doesn't help, however pressing `C-l` does help.  It
> seems that pressing any control character interactively fixes the
> issue.  But as soon as I use `feedkeys()` to automate the keypress,
> the fix stops working.

This very much sounds like the input method is not reset.
Try changing some of the 'im' options, such as 'iminsert' and
'imsearch'.

-- 
Imagine a world without hypothetical situations.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
I've tried all possible combinations of 'im' options (96):

    " source this code in Vim
    let counter = 0
    for imdisable in ['imdisable', 'noimdisable']
        for iminsert in ['iminsert=0', 'iminsert=1', 'iminsert=2']
            for imstyle in ['imstyle=0', 'imstyle=1']
                for imsearch in ['imsearch=-1', 'imsearch=0', 'imsearch=1', 'imsearch=2']
                    for imcmdline in ['noimcmdline', 'imcmdline']
                        let counter += 1
                        call writefile([printf('set %s %s %s %s %s | startinsert | ino <m-b> xxx', imdisable, iminsert, imstyle, imsearch, imcmdline)],
                            \ '/tmp/.test.'..printf('%02d', counter)..'.vim')
                    endfor
                endfor
            endfor
        endfor
    endfor

    # run this shell command
    $ for f in /tmp/.test.*.vim; do vim -Nu NONE -S "$f";done
    " press dead key '^' followed by 'a'

The result is always the same: Vim inserts `xxx` instead of `â`.  I tested without any code in `~/.Xresources`.

**Comment by brammool** [comment]
> I've tried all possible combinations of 'im' options (96):
> 
>     " source this code in Vim
>     let counter = 0
>     for imdisable in ['imdisable', 'noimdisable']
>         for iminsert in ['iminsert=0', 'iminsert=1', 'iminsert=2']
>             for imstyle in ['imstyle=0', 'imstyle=1']
>                 for imsearch in ['imsearch=-1', 'imsearch=0', 'imsearch=1', 'imsearch=2']
>                     for imcmdline in ['noimcmdline', 'imcmdline']
>                         let counter += 1
>                         call writefile([printf('set %s %s %s %s %s | startinsert | ino <m-b> xxx', imdisable, iminsert, imstyle, imsearch, imcmdline)],
>                             \ '/tmp/.test.'..printf('%02d', counter)..'.vim')
>                     endfor
>                 endfor
>             endfor
>         endfor
>     endfor
> 
>     # run this shell command
>     $ for f in /tmp/.test.*.vim; do vim -Nu NONE -S "$f";done
>     " press dead key '^' followed by 'a'
> 
> The result is always the same: Vim inserts `xxx` instead of `â`.

Hmm, I wonder if the method that makes the dead key work is not an input
method but some other mechanism.  You mentioned that typing any key with
CTRL makes it work.  Does it also work when only pressing CTRL, without
any other key?

-- 
It's totally unfair to suggest - as many have - that engineers are socially
inept.  Engineers simply have different objectives when it comes to social
interaction.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> Does it also work when only pressing CTRL, without any other key?

No, it doesn't work with only pressing `CTRL`.  I need to combine `CTRL` with another key.

**Comment by lacygoill** [comment]
No idea when it got fixed, but I can no longer reproduce, so I close the issue.
